### ISSUE TITLE
Cache regression with large guilds

### ISSUE THREAD

**Issue by jamesbt365** [serenity-rs/serenity#2528]
Large guilds (I have only tested this on a guild with more than 70,000 members) the guild may fail to cache properly. I have gotten this problem to appear at around 2pm EDT and carry over for a few hours due to an increased amount of data handling when the bot starts up.

I cannot get this issue to occur with the GUILD_PRESENCES intent disabled, because the large amount of data it sends.

![image](https://github.com/serenity-rs/serenity/assets/22662856/4b017cbd-d47c-4673-96d2-72664d481586)
This image above is a simple example of switching between f45b7e5e9f2b5ed593671972b25577ffca0e403a (working) and 5548af05f4234dc723b59217ef82bbf09b014068 (broken). The commits between these do not compile so the problem must be in that range of commits. Which you can see [here](https://github.com/serenity-rs/serenity/compare/8f3291370526205d6625476de2d660ab9a2965ee..5548af05f4234dc723b59217ef82bbf09b014068).

This is running on rust 1.72.0.

The code I used throughout all my tests is just the ping example with a tweak to the message event and with every intent enabled:
```rust
async fn message(&self, ctx: Context, msg: Message) {
    if let Some(guild_id) = msg.guild_id {
        println!("{:?}", msg.guild_id);
        if let Some(guild) = guild_id.to_guild_cached(&ctx) {
            let guild_name = guild.name.clone();
            println!("Message received in guild: {}", guild_name);
        }
    }
}
```

Thanks for everyone so far for helping me get to the conclusion of why this could occur. If you need any more information I'll try and provide it to you.

**Comment by jamesbt365** [comment]
I haven't been able to repro this since, but its school terms now so maybe I just am not getting as much data because not as many people are on the platform at once? Will investigate again soon.

**Comment by jamesbt365** [comment]
Can confirm it still happens.

**Comment by jamesbt365** [comment]
So this issue was actually masked by discord sometimes sending 0s in application_id on presence data causing GUILD_CREATE to not be parsed correctly, I guess I might just have been lucky on certain hours to not encounter this, this issue really was masked like hell, I should've just checked debug mode on in the first place. Oops!

Fixed in #2689 (next) and #2692 (current).
